src/hotspot/share/memory/iterator.inline.hpp

```diff
@@ -51,7 +51,11 @@ inline void ClaimMetadataVisitingOopIterateClosure::do_cld(ClassLoaderData* cld)
 
 inline void ClaimMetadataVisitingOopIterateClosure::do_klass(Klass* k) {
   ClassLoaderData* cld = k->class_loader_data();
-  ClaimMetadataVisitingOopIterateClosure::do_cld(cld);
+  if (cld != nullptr) {
+    ClaimMetadataVisitingOopIterateClosure::do_cld(cld);
+  } else {
+    assert_is_pending_aot_linked_class(k);
```

Comment:
Would there be a drawback to just adding `assert(AOTLinkedClassBulkLoader::is_pending_aot_linked_class(k), "sanity");` here and skip the changes the both .hpp and .cpp files?

Reply:
I changed the code to `assert(is_pending_aot_linked_class(k), "sanity")`. Since iterator.hpp is widely used, I avoid including aotLinkedClassBulkLoader.hpp in this header, and route the call through iterator.cpp.

Reply:
My proposal was to use it in this file (iterator.inline.hpp) and not iterator.hpp. I see no need to route this through the .cpp file.

Reply:
iterator.inline.hpp is included by many cpp file (580 out of 1409 .o files in hotspot), so I don't want to add aotLinkedClassBulkLoader.hpp to it.

```
scanning    580 iterator.inline.hpp
   2 found    580 oop.inline.hpp
   3 found    580 zIterator.inline.hpp
   4 found    580 zBarrierSet.inline.hpp
   5 found     60 g1OopClosures.inline.hpp
   6 found      7 g1FullGCOopClosures.inline.hpp
   7 found      5 psPromotionManager.inline.hpp
```

Reply:
From what I can tell it is a tiny header with very few dependencies. I don't think we gain much by trying to prevent it from being include. I would strongly prefer to have cleaner code in iterator.hpp/cpp than trying to get rid of that include.

Reply:
I have moved the aotLinkedClassBulkLoader.hpp to iterator.inline.hpp as you suggested:

https://github.com/openjdk/jdk/pull/21642/files#diff-9512406aa5bffbef807aa26007f54e2bfb6b0635e1829aa77bf483bdfd342c8cR30